core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -507,86 +518,126 @@ trait EntityComponent {
-      val entityQueries = entitiesToCopy.map { entity =>
-        findEntityByName(destWorkspaceContext.workspaceId, entity.entityType, entity.name).result.map {
-          case Seq() => None
-          case _ => Option(entity)
+      def getSoftConflicts(paths: Seq[Seq[AttributeEntityReference]]) = {
+        getCopyConflicts(destWorkspaceContext, paths.map(_.last)).map { conflicts =>
+          val conflictsAsRefs = conflicts.toSeq.map(_.toReference)
+          paths.filter(p => conflictsAsRefs.contains(p.last))
         }
       }
-      DBIO.sequence(entityQueries).map(_.toStream.collect { case Some(e) => e })
+
+      def buildSoftConflictTree(pathsRemaining: Seq[AttributeEntityReference], result: Seq[EntitySoftConflict]): Seq[EntitySoftConflict] = {
+        if(pathsRemaining.isEmpty) result
+        else Seq(EntitySoftConflict(pathsRemaining.head.entityType, pathsRemaining.head.entityName, buildSoftConflictTree(pathsRemaining.tail, result)))
+      }
+
+      val entitiesToCopyRefs = entityNames.map(name => AttributeEntityReference(entityType, name))
+
+      getHardConflicts(destWorkspaceContext.workspaceId, entitiesToCopyRefs).flatMap {
+        case Seq() => {
+          val pathsAndConflicts = for {
+            entityPaths <- getEntitySubtrees(sourceWorkspaceContext, entityType, entityNames)
+            softConflicts <- getSoftConflicts(entityPaths)
+          } yield (entityPaths, softConflicts)
+
+          pathsAndConflicts.flatMap { case (entityPaths, softConflicts) =>
+            if(softConflicts.isEmpty || linkExistingEntities) {
+              val allEntityRefs = entityPaths.flatten
+              val allConflictRefs = softConflicts.flatten
+              val entitiesToClone = allEntityRefs diff allConflictRefs
+
+              for {
+                entities <- list(sourceWorkspaceContext, entitiesToClone)
+                _ <- cloneEntities(destWorkspaceContext, entities.toSet, allConflictRefs)
+              } yield EntityCopyResponse(entities.map(_.toReference).toSeq, Seq.empty, Seq.empty)
+            } else {
+              val unmergedSoftConflicts = softConflicts.flatMap(c => buildSoftConflictTree(c, Seq.empty))
+              val grouped = unmergedSoftConflicts.groupBy(c => (c.entityType, c.entityName))
+              DBIO.successful(EntityCopyResponse(Seq.empty, Seq.empty, grouped.map{ case ((entityType, entityName), conflicts) => EntitySoftConflict(entityType, entityName, conflicts.flatMap(_.conflicts))}.toSeq))
+            }
+          }
+        }
+        case hardConflicts => DBIO.successful(EntityCopyResponse(Seq.empty, hardConflicts.map(c => EntityHardConflict(c.entityType, c.entityName)), Seq.empty))
+      }
+    }
+
+    def getCopyConflicts(destWorkspaceContext: SlickWorkspaceContext, entitiesToCopy: TraversableOnce[AttributeEntityReference]): ReadAction[TraversableOnce[EntityRecord]] = {
```

Comment:
I think that the name is accurate. It returns the subset of `entitiesToCopy` that exist in `destWorkspaceContext`. 

The changes to this are just a performance improvement on what was already there. Previously it would look up each entity in full one by one. The only information we need from this method is which entities in `entitiesToCopy` exist and we can do that in one lightweight query instead of `n` full attribute joins